### PR TITLE
[RISCV][GISEL] Legalize G_ZEXT, G_SEXT, G_ANYEXT, G_SPLAT_VECTOR, and G_ICMP

### PR DESCRIPTION
**Pull request by michaelmaitland**
This PR contains a series of multiple commits which I did not break up into their own separate PRs since there are cyclical dependencies between them. I explain these dependencies below.

14693ad35293006b2682004e3783e01e52ce9aab does legalization for G_ZEXT, G_SEXT, and G_ANYEXT with scalable vector type.

However, legalization of s1 element vectors for G_ZEXT, G_SEXT, and G_ANYEXT build G_SPLAT_VECTOR and G_SELECT instructions. As a result, we legalize G_SPLAT_VECTOR in 3045357ab146578321159da3ff7e83e0cf545c71 and G_SELECT in #85540.

Legalization of G_SPLAT_VECTOR relies on building a G_ICMP where the true and false values are s1. As a result, we include 5608ef326a1b9dd407070dd3d959cd129f92de8e to do legalization for G_ICMP.

I wanted to have G_ICMP work as its own PR, but since the true and false values are s1 element vectors, the legalization depends on sign extension, creating a cyclic dependency between these patches.

### CODE REVIEW

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 1]
```diff
@@ -374,6 +379,11 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
       .clampScalar(0, s32, sXLen)
       .lowerForCartesianProduct({s32, sXLen, p0}, {p0});
 
+  getActionDefinitionsBuilder(G_SPLAT_VECTOR)
+      .legalIf(all(typeIsLegalIntOrFPVec(0, IntOrFPVecTys, ST),
+                   typeInSet(1, {s8, s16, s32, sXLen})))
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:384]: This should only be sXLen. There are no isel patterns with any other type.

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 2]
```diff
@@ -495,6 +505,124 @@ bool RISCVLegalizerInfo::shouldBeInConstantPool(APInt APImm,
...
+
+  MachineRegisterInfo &MRI = *MIB.getMRI();
+  Register Dst = MI.getOperand(0).getReg();
+  Register Src = MI.getOperand(1).getReg();
+
+  LLT DstTy = MRI.getType(Dst);
+  LLT SrcTy = MRI.getType(Src);
+
+  // The only custom legalization of extends we handle are vector extends.
+  if (!DstTy.isVector() || !SrcTy.isVector())
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:527]: Isn't this controlled by the `customIf` in the legalizer rules? Why would a scalar type get here?

File: llvm/lib/CodeGen/RegisterBankInfo.cpp [hunk 3]
```diff
@@ -484,9 +484,10 @@ void RegisterBankInfo::applyDefaultMapping(const OperandsMapper &OpdMapper) {
       // the storage. However, right now we don't necessarily bump all
       // the types to storage size. For instance, we can consider
       // s16 G_AND legal whereas the storage size is going to be 32.
-      assert(OrigTy.getSizeInBits() <= NewTy.getSizeInBits() &&
-             "Types with difference size cannot be handled by the default "
-             "mapping");
+      assert(
+          TypeSize::isKnownLE(OrigTy.getSizeInBits(), NewTy.getSizeInBits()) &&
+          "Types with difference size cannot be handled by the default "
+          "mapping");
```
Comment (arsenm) [inline llvm/lib/CodeGen/RegisterBankInfo.cpp:490]: Pre-commit all these generic getNumElements->getElementCount changes?

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 4]
```diff
@@ -141,6 +141,9 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
 
   if (ST.is64Bit()) {
     getActionDefinitionsBuilder({G_ZEXT, G_SEXT, G_ANYEXT})
+        .customIf(typeIsLegalBoolVec(1, BoolVecTys, ST))
+        .legalIf(all(typeIsLegalIntOrFPVec(0, IntOrFPVecTys, ST),
+                     typeIsLegalIntOrFPVec(1, IntOrFPVecTys, ST)))
```
Comment (arsenm) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:146]: legal operations should be preferred as the first action

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 5]
```diff
@@ -149,7 +152,11 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
         .maxScalar(0, sXLen)
         .lower();
   } else {
-    getActionDefinitionsBuilder({G_ZEXT, G_SEXT, G_ANYEXT}).maxScalar(0, sXLen);
+    getActionDefinitionsBuilder({G_ZEXT, G_SEXT, G_ANYEXT})
+        .customIf(typeIsLegalBoolVec(1, BoolVecTys, ST))
```
Comment (arsenm) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:156]: legal operations should be preferred as the first action

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 6]
```diff
@@ -229,6 +236,9 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
 
   getActionDefinitionsBuilder(G_ICMP)
       .legalFor({{sXLen, sXLen}, {sXLen, p0}})
+      .widenScalarOrEltToNextPow2OrMinSize(1, 8)
+      .legalIf(all(typeIsLegalBoolVec(0, BoolVecTys, ST),
```
Comment (arsenm) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:240]: legal operations should be preferred as the first action

File: llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp [hunk 7]
```diff
@@ -1713,6 +1713,15 @@ LegalizerHelper::LegalizeResult LegalizerHelper::narrowScalar(MachineInstr &MI,
     MI.eraseFromParent();
     return Legalized;
   }
+  case TargetOpcode::G_SPLAT_VECTOR: {
+    if (TypeIdx != 1)
+      return UnableToLegalize;
+
+    Observer.changingInstr(MI);
+    narrowScalarSrc(MI, NarrowTy, 1);
```
Comment (topperc) [inline llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:1721]: This silently drops the upper bits of the scalar and creates a splat vector where the source scalar type is smaller than the vector element type.

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 8]
```diff
@@ -229,6 +236,9 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
 
   getActionDefinitionsBuilder(G_ICMP)
       .legalFor({{sXLen, sXLen}, {sXLen, p0}})
+      .legalIf(all(typeIsLegalBoolVec(0, BoolVecTys, ST),
+                   typeIsLegalIntOrFPVec(1, IntOrFPVecTys, ST)))
+      .widenScalarOrEltToNextPow2OrMinSize(1, 8)
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:241]: Does this make the `.widenScalarToNextPow2(1)` on the next line redundant?

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 9]
```diff
@@ -570,7 +591,146 @@ bool RISCVLegalizerInfo::legalizeVScale(MachineInstr &MI,
...
+
+  Register Dst = MI.getOperand(0).getReg();
+  Register SplatVal = MI.getOperand(1).getReg();
+
+  LLT VecTy = MRI.getType(Dst);
+  LLT XLenTy(STI.getXLenVT());
+
+  // Handle case of s64 element vectors on rv32
+  if (XLenTy.getSizeInBits() == 32 &&
+      VecTy.getElementType().getSizeInBits() == 64) {
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:700]: This probably does the wrong thing for `f64` on RV32?

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp (hunk 8 above)
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:241]: With `widenScalarToNextPow2` if the size was s1, then it would not be widened. There is nothing in the MachineVerifier that says it couldn't be an i1 scalar.

Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:241]: Doesn't MinSize there mean widen to next power of 2 that is >= 8?

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:240]: The description says

> Widen the scalar to the next power of two that is at least MinSize. No effect if the type is not a scalar or is a power of two.

I took

> no effect if the type is not a ... power of two

to mean that exactly that.

I think you took it as

> no effect if the type is not a .. power of two or at least MinSize

I will check how it actually works.

Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:240]: Please fix the comment(as a separate patch) once you figure it out. The version without `OrMinSize` suffix has the same description.

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:240]: Removed the redundancy here and pushed an NFC commit fixing the docstring in 1d5e5f4d3c68e63ced47ee9b17d62fb995aa1e62

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp (hunk 9 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:698]: If s64 is legal from D extensions and s64 vectors are legal, then I think we should keep the G_SPLAT_VECTOR as legal. Naively, this would cause us to go from GPR->mem->FPR->VFMV_VF which isn't efficient, but is correct.

I believe this would show up as (G_SPLAT_VECTOR (G_MERGE_VALUES Lo, Hi)), we could potentially add a post-legalize combine to turn it into G_SPLAT_VECTOR_SPLIT_I64_VL or just special case it in RISCVInstructionSelector.

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp]: I have updated this patch to legalize as described. I plan to post regbankselect and selection in another PR.

**Comment by michaelmaitland** [comment]
I removed regbank selection and instruction selection from this patch to make the review smaller. I will post those as their own PRs.

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 10]
```diff
@@ -412,6 +427,27 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
...
+                       typeIs(1, sXLen)))
+          .customIf(all(typeIsLegalBoolVec(0, BoolVecTys, ST), typeIs(1, s1)));
+  // Handle case of s64 element vectors on RV32. If the subtarget does not have
+  // f64, then try to lower it to G_SPLAT_VECTOR_SPLIT_64_VL. If the subtarget
+  // does have f64, then we don't know whether the type is an f64 or an i64,
+  // so mark the G_SPLAT_VECTOR as legal and decide later what to do with it,
+  // depending on how the instructions it consumes are legalized. They are not
+  // legalized yet since legalization is in reverse postorder, so we cannot
+  // make the decision at this moment.
+  if (XLen == 32)
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:442]:
```suggestion
  if (XLen == 32) {
    if (ST.hasVInstructionF64() && ST.hasStdExtD())
      SplatActions
          .legalIf(all(typeInSet(0, {nxv1s64, nxv2s64, nxv4s64, nxv8s64}),
                       typeIs(1, s64)))
    else if (ST.hasVInstructionsI64())
        SplatActions.customIf(all(typeInSet(0, {nxv1s64, nxv2s64, nxv4s64, nxv8s64}),
                      typeIs(1, s64)));
    }
```

File: llvm/lib/Target/RISCV/RISCVInstrGISel.td [hunk 11]
```diff
@@ -32,3 +32,28 @@ def G_READ_VLENB : RISCVGenericInstruction {
...
+// Pseudo equivalent to a RISCVISD::VMSET_VL
+def G_VMSET_VL : RISCVGenericInstruction {
+  let OutOperandList = (outs type0:$dst);
+  let InOperandList = (ins type1:$vl);
+  let hasSideEffects = false;
+}
+def : GINodeEquiv<G_VMSET_VL, riscv_vmset_vl>;
+
+// Pseudo equivalent to a RISCVISD::SPLAT_VECTOR_SPLIT_I64_VL. There is no
+// record to mark is equivalent to using GINodeEquiv because it gets lowered
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrGISel.td:53]: `is equivalent` -> `as equivalent`?

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/rvv/legalize-anyext.mir [hunk 12]
```diff
@@ -0,0 +1,1011 @@
...
+
+# Extend from s1 element vectors
+---
+name:            anyext_nxv1i8_nxv1i1
+legalized:       false
+tracksRegLiveness: true
+body:             |
+  bb.0.entry:
+    ; RV32-LABEL: name: anyext_nxv1i8_nxv1i1
+    ; RV32: [[DEF:%[0-9]+]]:_(<vscale x 1 x s8>) = G_IMPLICIT_DEF
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/legalizer/rvv/legalize-anyext.mir:13]: This is bad test case. Don't use G_IMPLICIT_DEF as the input.

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/rvv/legalize-xor.mir [hunk 13]
```diff
@@ -9,8 +9,8 @@ body:             |
     ; CHECK-LABEL: name: test_nxv1i8
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/legalizer/rvv/legalize-xor.mir:9]: This seems unrelated to this patch?

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 14]
```diff
@@ -576,7 +602,146 @@ bool RISCVLegalizerInfo::legalizeVScale(MachineInstr &MI,
...
+  assert(Opc == TargetOpcode::G_ZEXT || Opc == TargetOpcode::G_SEXT ||
+         Opc == TargetOpcode::G_ANYEXT);
+
+  MachineRegisterInfo &MRI = *MIB.getMRI();
+  Register Dst = MI.getOperand(0).getReg();
+  Register Src = MI.getOperand(1).getReg();
+
+  LLT DstTy = MRI.getType(Dst);
+  int64_t ExtTrueVal =
+      Opc == TargetOpcode::G_ZEXT || Opc == TargetOpcode::G_ANYEXT ? 1 : -1;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:626]: Wouldn't it be simpler to check G_SEXT here?

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/rvv/legalize-anyext.mir [hunk 15]
```diff
@@ -0,0 +1,1589 @@
...
+    ; RV64-NEXT: [[C:%[0-9]+]]:_(s32) = G_CONSTANT i32 0
+    ; RV64-NEXT: [[ANYEXT:%[0-9]+]]:_(s64) = G_ANYEXT [[C]](s32)
+    ; RV64-NEXT: [[SPLAT_VECTOR:%[0-9]+]]:_(<vscale x 1 x s8>) = G_SPLAT_VECTOR [[ANYEXT]](s64)
+    ; RV64-NEXT: [[C1:%[0-9]+]]:_(s32) = G_CONSTANT i32 1
+    ; RV64-NEXT: [[ANYEXT1:%[0-9]+]]:_(s64) = G_ANYEXT [[C1]](s32)
+    ; RV64-NEXT: [[SPLAT_VECTOR1:%[0-9]+]]:_(<vscale x 1 x s8>) = G_SPLAT_VECTOR [[ANYEXT1]](s64)
+    ; RV64-NEXT: [[SELECT:%[0-9]+]]:_(<vscale x 1 x s8>) = G_SELECT [[COPY]](<vscale x 1 x s1>), [[SPLAT_VECTOR1]], [[SPLAT_VECTOR]]
+    ; RV64-NEXT: $v8 = COPY [[SELECT]](<vscale x 1 x s8>)
+    ; RV64-NEXT: PseudoRET implicit $v8
+    %1:_(<vscale x 1 x s1>) = COPY $v8
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/legalizer/rvv/legalize-anyext.mir:38]: This is not the correct ABI for passing a mask into a function. Mask should use v0. Is this handcrafted?

Comment (michaelmaitland) [reply llvm/test/CodeGen/RISCV/GlobalISel/legalizer/rvv/legalize-anyext.mir:38]: hand crafted. I can update to v0.

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/rvv/legalize-xor.mir (hunk 13 above)
Comment (topperc) [reply llvm/test/CodeGen/RISCV/GlobalISel/legalizer/rvv/legalize-xor.mir:8]: Not addressed. Should this change be committed as its own NFC?

Comment (michaelmaitland) [reply llvm/test/CodeGen/RISCV/GlobalISel/legalizer/rvv/legalize-xor.mir:8]: I'll commit it as its own NFC. It keeps making its way in again. I've removed it 2x since you commented here!

**Review by topperc — APPROVED** [review]
LGTM

**Comment by michaelmaitland** [comment]
Closed by 8aa3a77eaf198afb7e01453e6daf6566b687945d
